Applications/Utils/ModelPreparation/PartitionMesh/MeshPartitioning.cpp

```diff
@@ -0,0 +1,493 @@
+                if (ghost_counter == 0)
+                {
+                    partition_regular_elements.push_back(elem);
+                }
+                else // ghost element
+                {
+                    partition_ghost_elements.push_back(elem);
+                    elems_non_ghost_nodes_local_ids[elem->getID()]
+                        = std::move(nonghost_nodes_local_ids);
+                }
+
+                elems_reseved[elem->getID()] = true;
+            }
+        }
+
+        // Add ghost nodes in ghost elements to the node vector of this partition
+        // nodes_reseved is reused without cleaning
+        // Mark the non-ghost nodes for each ghost element
+        for (const auto ghost_elem : partition_ghost_elements)
+        {
+            for (unsigned k=0; k<ghost_elem->getNNodes(); k++)
+                nodes_reseved[ghost_elem->getNodeIndex(k)] = false;
+
+            // Mark non-ghost nodes
+            std::vector<unsigned>& local_node_ids = elems_non_ghost_nodes_local_ids[ghost_elem->getID()];
+            for (std::size_t k=0; k<local_node_ids.size(); k++)
+                nodes_reseved[ghost_elem->getNodeIndex(local_node_ids[k])] = true;
+        }
+        // Add the ghost nodes to the node vector of this partition
+        for (const auto ghost_elem : partition_ghost_elements)
+        {
+            for (unsigned k=0; k<ghost_elem->getNNodes(); k++)
+            {
+                const unsigned node_id = ghost_elem->getNodeIndex(k);
+                if (nodes_reseved[node_id])
+                    continue;
+                nodes_reseved[node_id] = true;
+                partition_nodes.push_back(ghost_elem->getNode(k));
+            }
+        }
+
+        partition_end_node_id[ipart] = partition_nodes.size();
+        // Renumber
+        PetscInt local_id = 0;
+
+        std::vector<unsigned> nodes_local_ids_partition;
+        nodes_local_ids_partition.reserve(nnodes);
+        for (std::size_t i=0; i<nnodes; i++)
+        {
+            nodes_local_ids_partition.push_back(0);
+        }
+
+        for (std::size_t i = partition_start_node_id[ipart];
+                i < end_non_ghost_node_id[ipart]; i++)
+        {
+            const unsigned node_id = partition_nodes[i]->getID();
+            node_global_ids[node_id] = global_node_id_offset;
+            nodes_local_ids_partition[node_id] = local_id;
+            local_id++;
+            global_node_id_offset++;
+        }
+        // Assign local IDs to ghost nodes
+        for (std::size_t i = end_non_ghost_node_id[ipart];
+                i < partition_end_node_id[ipart]; i++)
+        {
+            nodes_local_ids_partition[partition_nodes[i]->getID()] = local_id;
+            local_id++;
+        }
+
+        // Count the number of integer variables used to define non-ghost elements
+        const PetscInt num_non_ghost_elems = partition_regular_elements.size();
+        PetscInt nmb_element_idxs = 3 * num_non_ghost_elems;
```

Comment:
The total number of the integers in an element vector is
 3 * vector size (offset, element type, material ID )  + sum (number of nodes of each element).
Has added a comment for this
https://github.com/wenqing/ogs/blob/b1f71428a0883f497305695b8a29716b5ec1c64e/Applications/Utils/ModelPreparation/PartitionMesh/NodeWiseMeshPartitioner.h#L112